### PR TITLE
Add aml_pipeline

### PR DESCRIPTION
**Pull request by orientnab**

### CODE REVIEW

File: docs/rst/tutorials/aml_pipeline.rst [hunk 1]
```diff
@@ -1,24 +1,452 @@
+.. include:: /rst/exports/roles.include
+
 .. _tutorials_aml_pipeline:
 
-############
 AML Pipeline
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: I'd remove this title and replace it with "Running a binary classifier with MNIST". The tutorial does not have to be name "AML Pipeline" as this section was just created to build the skeleton of the documentation. Also, change the name of the file to ``binary_classifier_mnist.rst``.

File: docs/rst/tutorials/aml_pipeline.rst [hunk 2]
```diff
@@ -1,24 +1,452 @@
...
 ############
 
-.. warning::
+You can request this Jupyter notebook and AML libraries in :ref:`Get access to AML Toolkit <get_access_toolkit>`.
+
+
+Running a binary classifer with MNIST
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]:
```suggestion
Running a binary classifier with MNIST
```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 3]
```diff
@@ -1,24 +1,452 @@
...
+*************************************
+
+Load AML libraries
+==================
+
+Most functionality to work with AML can be found in two packages:
+- ``aml_engine.amlSimpleLibrary`` contains functions and classes to create and manipulate AML models,
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: The list is not rendered correctly. Add a line break.

```suggestion

- ``aml_engine.amlSimpleLibrary`` contains functions and classes to create and manipulate AML models.
```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 4]
```diff
@@ -1,24 +1,452 @@
...
+
+- ``t10k-images-idx3-ubyte``
+- ``t10k-labels-idx1-ubyte``
+- ``train-images-idx3-ubyte``
+- ``train-labels-idx1-ubyte``
+
+The dataset can be downloaded from:
+http://yann.lecun.com/exdb/mnist/
+and
+https://huggingface.co/datasets/ylecun/mnist
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]:
```suggestion
The dataset can be downloaded from `Yann Lecun MNIST database <http://yann.lecun.com/exdb/mnist/>`__ or `Huggingface <https://huggingface.co/datasets/ylecun/mnist>`__.
```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 5]
```diff
@@ -1,24 +1,452 @@
...
+.. code-block:: python
+
+    from MNIST import mnistGenerator
+
+The ``mnistGenerator`` searches and loads the datasets and provides functionality to load individual digits.
+
+Embedding constants
+===================
+
+A simple embedding for black and white images:
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: Same as before, add a break line before lists to render them correctly. 

```suggestion
A simple embedding for black and white images:

```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 6]
```diff
@@ -1,24 +1,452 @@
...
+- constants for each black pixel `B(i,j)` (height x width),
+- constants for each white pixel `W(i,j)` (height x width),
+- constants for each label (10 if using all digits).
+
+For instance, with 2x2 images the embedding set would be
+``[B(1,1), B(1,2), B(2,1), B(2,2), W(1,1), W(1,2), W(2,1), W(2,2), label(1), label(2), ...]``
+
+### Embedding function
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]:
```suggestion
Embedding function
==================
```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 7]
```diff
@@ -1,24 +1,452 @@
...
+
+    balance = 9
+    pBatchSize = 100
+    nBatchSize = 100 * balance
+    maxPTrainingExamples = 6000
+    maxNTrainingExamples = 54000
+
+Notice that we are picking one target digit.
+The model will discriminate between that digit (positive class), and the rest.
+It is also possible to slightly modify this script for the multiclass problem.
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: I'd move the description before the code.

File: docs/rst/tutorials/aml_pipeline.rst [hunk 8]
```diff
@@ -1,24 +1,452 @@
...
+
+Model and batch trainer initialisation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Initialise the ``model`` object that contains:
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: Line break

```suggestion
Initialise the ``model`` object that contains:

```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 9]
```diff
@@ -1,24 +1,452 @@
...
+    fraction of negative indicators:  0.14285714285714285 , union model fraction:  15
+    BATCH: 948 batchSize( 100 , 900              ------------- master:  model size 9977
+
+    ... following iterations ...
+
+
+Postprocessing
+==============
+
+Here you can explore the atomization in the last model: `model.atomization`
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]:
```suggestion
Here you can explore the atomization in the last model: `model.atomization`.
```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 10]
```diff
@@ -1,24 +1,452 @@
...
+
+But also the cumulative model that has been built combining the atomizations across the whole training process: `batchLearner.unionModel`
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]:
```suggestion
But also the cumulative model that has been built combining the atomizations across the whole training process: `batchLearner.unionModel`.
```

File: docs/rst/tutorials/aml_pipeline.rst [hunk 11]
```diff
@@ -1,24 +1,452 @@
...
-Prerequisites
-*************
+As an example, below we show 100 atoms in ``model.atomization``.
+Black and white pixels indicate that the constant is present in the atom; red, that both constants are present; and gray, that none is present.
 
-.. note::
+.. image:: /rst/figures/tutorials/example_atoms_mnist.png
+   :width: 400
+   :alt: A hundred atoms from the atomization of MNIST.
+   :align: center
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: I'm missing another section for inference, I mean, given an image from MNIST dataset not used for training, infer if it belongs to that class or not.

File: docs/rst/tutorials/aml_pipeline.rst [hunk 12]
```diff
@@ -1,24 +1,452 @@
+.. include:: /rst/exports/roles.include
+
 .. _tutorials_aml_pipeline:
 
-############
 AML Pipeline
 ############
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: Add a more detailed description of this tutorial. Something like "This tutorial covers...".

File: docs/rst/tutorials/aml_pipeline.rst (hunk 7 above)
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: Move this sentence to another section at the end of this tutorial called "Next Steps" and add a link there to another tutorial named "Extending the MNIST binary classifier to a multiclass classifier". Create the file for this tutorial and add it to the index. Then on the file just add the title and a warning with a "Work in progress" statement.

File: docs/rst/tutorials/aml_pipeline.rst [hunk 13]
```diff
@@ -1,24 +1,452 @@
...
+    import random
+    for at in random.choices(model.atomization, k=10):
+        print(at.ucs)
+
+    batchLearner.unionModel.sort(reverse=True,key=lambda at: len(at.ucs))
+    print(f"{cGreen}Largest atom{cReset}")
+    print(batchLearner.unionModel[0].ucs)
+
+
+This displays the following information::
```
Comment (rsanchez15) [inline docs/rst/tutorials/aml_pipeline.rst]: I'd explain more about this information, i.e. what can we deduce from it or how could we interpret it to check if everything looks good.

**Review by rsanchez15 — CHANGES_REQUESTED** [review]
Upload the python files for this tutorial to another directory in resources folder. Also, add a link to this directory at the beginning of the tutorial.

File: docs/rst/tutorials/aml_pipeline.rst (hunk 7 above)
Comment (orientnab) [reply docs/rst/tutorials/aml_pipeline.rst]: No sé si es una buena idea. Es sencillo extender el código a multiclase, pero la inferencia se vuelve mucho más complicada (a nivel estadístico, no de AML)

Comment (rsanchez15) [reply docs/rst/tutorials/aml_pipeline.rst]: Let's skip it for now then

**Review by rsanchez15 — APPROVED** [review]

**Pull request merged**